### DISCUSSION TITLE
Fullscreen performance

### DISCUSSION THREAD

**Discussion by JustCauseWhyNot** [bakkeby/patches#55] [Q&A] [answered]
When using the [fullscreen](https://github.com/bakkeby/patches/wiki/togglefullscreen) patch to turn a windowed window into a fullscreen window does it still have the performance of a windowed window or the higher performance of a fullscreen window?

**Comment by bakkeby** [comment, marked as answer]
I felt this was better placed under discussions rather than issues.

I do not know the exact answer to this question, but I think that it depends on what the actual cause of the performance issues is when running the game in windowed mode.

For all practical purposes you have an X window and a fullscreen X window merely takes up the entire screen and is placed on top of all other windows (including the bar etc.).

From the window manager's side, and the X server's side for that matter, there is no practical difference of running a game in fullscreen compared to running Firefox in fullscreen. As such I would assume that if there are performance issues then they may still be there when running a game in fullscreen in dwm.

It is not an exclusive fullscreen like you may be used to from Windows XP, as in keybindings to switch tags, etc. will still work. That said some games, or the programs managing said games, may change the resolution of the screen when going fullscreen which may potentially improve performance.

If you are using a compositor then that can also have an impact on performance. In picom for example there is an option to stop handling transparency effects and the like when an opaque fullscreen window has been detected.

```tcl
# Unredirect all windows if a full-screen opaque window is detected,
# to maximize performance for full-screen windows. Known to cause flickering
# when redirecting/unredirecting windows.
#
unredir-if-possible = true;
```

@vide0hanz has more experience than me in this area and may have something to add.

**Reply by JustCauseWhyNot** [reply]
> I felt this was better placed under discussions rather than issues. 

I agree

> If you are using a compositor

I currently don't use a compositor, but this is good to know if I do end up using one.
> It is not an exclusive fullscreen like you may be used to from Windows XP, as in keybindings to switch tags, etc. will still work. That said some games, or the programs managing said games, may change the resolution of the screen when going fullscreen which may potentially improve performance.

It's not so much performance issues as it is slight performance loss. But the resolution is the same either way. The game I'm referring too is fallout nv, and using this [performance guide](https://wallsogb.github.io/FalloutNV-Performance-Guide/#DisplayModeComparison). Because I currently have an issue with it were if I change focus off of the window while the game is fullscreened by in game settings, and focus back it shrinks into top left corner of screen. If I run in windowed mode, and use your goated patches then it's not an issue. But I'm trying to limit any performance loss, and increased input latency that comes with running the game in windowed mode.

Also is there a way too launch the window into fullscreen mode with patches, and also have it open in a separate tag/workspace?

Thank you for your help.

**Reply by bakkeby** [reply]
> Also is there a way too launch the window into fullscreen mode with patches, and also have it open in a separate tag/workspace?

In _general_ this should not be an issue, but...

> Because I currently have an issue with it were if I change focus off of the window while the game is fullscreened by in game settings, and focus back it shrinks into top left corner of screen.

One thing that we have learnt is that using the [awesomebar](https://dwm.suckless.org/patches/awesomebar/) patch for dwm and minimising (hiding) the window and revealing it again often fixes these kind of issues (including the black screen issue).

The reason for this, as far as I have been able to tell, is that the game ends up in a different state (e.g. minimised) when it loses input focus, but this is not actually recorded or shown even if you try to pull data for the window using `xprop` or `xwininfo`. Explicitly setting the client state to NormalState when giving input focus back solves many, but not all, of these kind of issues.

https://github.com/bakkeby/patches/blob/2736bc046d5ceef55bfc283b4543eae5d37a1d51/dwm/dwm-steam-6.3.diff#L61-L69

The issues vary depending on the game as well as what game launcher is used, e.g. Steam, Lutris and wine may all behave differently. It is not clear whether proton has any part in this.

Unrelated but it is also possible to run any .exe under proton within the need to launch the Steam client using a script \[[ref](https://gist.github.com/brunoais/575db9912368124d3223784afe20158c)\] \[[ref](https://gist.github.com/thingsiplay/3a933f557277906dc6b0e03ec8df5dbd)\]. It is also supposed to be possible to install the version of proton that Valve uses, which may have better compatibility (but I'm only guessing as I haven't fiddled with it).

**Reply by JustCauseWhyNot** [reply]
>One thing that we have learnt is that using the [awesomebar](https://dwm.suckless.org/patches/awesomebar/) patch for dwm and minimising (hiding) the window and revealing it again often fixes these kind of issues (including the black screen issue).

Is there another way to do that without using the awesomebar patch. Not a huge issue, because I'd install the patch if need be.
>The issues vary depending on the game as well as what game launcher is used, e.g. Steam, Lutris and wine may all behave differently. It is not clear whether proton has any part in this.

The problematic program is being run through [wine-ge](https://github.com/GloriousEggroll/wine-ge-custom). They way I launch everything is using a [desktop file](https://gist.github.com/5c64b3ea0e062bc0a9d087cdadac67c6) to launch mod organizer 2(mod manager for fallout nv), and from there I launch fallout nv through the manager. Would the steam patch you listed potentially apply to what I'm doing or would I need to adapt it?

**Reply by JustCauseWhyNot** [reply]
>One thing that we have learnt is that using the [awesomebar](https://dwm.suckless.org/patches/awesomebar/) patch for dwm and minimising (hiding) the window and revealing it again often fixes these kind of issues (including the black screen issue).

I can say that this does bring the game back, but I have what looks like a theater view with two black bars on top, and bottom of screen.

>Unrelated but it is also possible to run any .exe under proton within the need to launch the Steam client using a script [[ref](https://gist.github.com/brunoais/575db9912368124d3223784afe20158c)] [[ref](https://gist.github.com/thingsiplay/3a933f557277906dc6b0e03ec8df5dbd)]. It is also supposed to be possible to install the version of proton that Valve uses, which may have better compatibility (but I'm only guessing as I haven't fiddled with it).

If I ran the game through proton would the steam patch come into effect then? That way I wouldn't need to use awesome bar.

**Reply by bakkeby** [reply]
What I meant is that just adding that one line to call setclientstate to NormalState at the end of the setfocus function may solve your problem (without the need for the awesomebar patch).

**Reply by JustCauseWhyNot** [reply]
Well I've applied both patches. Without using awesome's keybinds then the problem still persisted. But if I hid, and then unhid the window it would work again.

**Reply by JustCauseWhyNot** [reply]
> 
> https://github.com/bakkeby/patches/blob/2736bc046d5ceef55bfc283b4543eae5d37a1d51/dwm/dwm-steam-6.3.diff#L61-L69
> 

Would it be possible to adapt to work with more than just steam windows. Such as any window, or specified game launcher?

**Reply by bakkeby** [reply]
I put together this (more generic) patch which aims to automatically minimize games running in fullscreen when moving to another tag and to unminimize them automatically when moving back.
https://github.com/bakkeby/patches/blob/master/dwm/dwm-games-6.3.diff

This will likely conflict with the awesombar patch mind you. I didn't include the configurerequest restrictions from the steam patch as that is something that can be looked into separately if need be.

If you can test this then that would be appreciated.

**Reply by JustCauseWhyNot** [reply]
I certainly will test this. I've already removed awesomebar patch as I was only testing it for troubleshooting purposes. But I'm going too bed I'll test in 6ish hrs from now.

**Reply by JustCauseWhyNot** [reply]
I tested it, and it didn't work.

**Reply by JustCauseWhyNot** [reply]
I'm dumb. I didn't set window tags properly. I can't say if it's worked or not yet.

**Reply by JustCauseWhyNot** [reply]
After setting things correctly it sorta doesn't work. The main thing is if I switch focus away, and then back it's just a black screen. Here's a [screenshot](https://drive.google.com/file/d/1bqw8GT-Gri9rGQMbvbn492pwClK1Z57p/view?usp=sharing). It feels like it's about to work, but it just doesn't.

**Reply by vide0hanz** [reply]
Just chiming in here. In my experience there is no performance loss when forcing a game window into fullscreen via @bakkeby 's patch. I can also confirm that the aforementioned picom configuration setting works as intended.

I tested this by enabling adaptive sync in my Nvidia settings and displaying the g-sync indicator and FPS counter. Every time I placed a game in fullscreen via this patch, adaptive sync would kick on as expected. Adaptive sync on Xorg only works when certain conditions are met; mainly that there is only one monitor enabled in the current X Screen and the application is in "true" fullscreen.

I'm bringing this up because your question implies that you are unsure whether or not toggling fullscreen is actually fullscreen rather than a borderless window. In my experience there is no distinction between the two as far as Xorg is concerned.

**Reply by JustCauseWhyNot** [reply]
I've benchmarked both windowed forced into fullscreen, and "native" fullscreen. There's a significant difference in the 1%, and 0.1% lows it's just very choppy.

**Reply by JustCauseWhyNot** [reply]
It's actually working now. I don't know what changed, but it's working now. Probably restarting, or something else I did. Thank you very much for your help.

**Reply by JustCauseWhyNot** [reply]
I found out why I didn't think it worked. It's because it doesn't if I immediately send it to another tag before the program actually loads anything onto my screen.

**Reply by bakkeby** [reply]
Something else that I haven't been able to confirm is whether the game should also be taken out of fullscreen when losing focus, and be brought back into fullscreen when receiving focus again.

**Reply by JustCauseWhyNot** [reply]
Interesting. All I know is currently it works flawlessly as long as I let the game load to main menu before moving it around and changing fous.

**Reply by vide0hanz** [reply]
> Something else that I haven't been able to confirm is whether the game should also be taken out of fullscreen when losing focus, and be brought back into fullscreen when receiving focus again.

I believe this is down to personal preference. I often use the losefullscreen patch to force fullscreen applications into windowed mode. Back when I was playing around with manipulating game windows vs Steam client windows I found it very intuitive for games to be brought back into fullscreen when receiving focus. Some people prefer to play certain games in windowed mode, however.

**Reply by JustCauseWhyNot** [reply]
> Something else that I haven't been able to confirm is whether the game should also be taken out of fullscreen when losing focus, and be brought back into fullscreen when receiving focus again.

Could you explain what the difference is from the current behavior of the patch you've provided? Because if it's what I think it is then I think it'd be a good idea too implement that. Because when I change focus too a separate window the game still is receiving input, and that doesn't seem right.

**Reply by JustCauseWhyNot** [reply]
I can also so I just tried out the losefullscreen patch, and it broke the "alttab" functionality of fnv.

**Comment by JustCauseWhyNot** [comment]
I've got another question. Does it make sense for the sensitivity of a game to get messed up when changing focus away from it? I'm talking about fnv, and when I switch away from it & back the menu sensitivity is much faster.

**Reply by bakkeby** [reply]
I didn't get back to you on this. I'm not sure if it "makes sense" for a game to behave this way when losing and gaining focus, but this can evidently happen. I am not entirely sure what the mechanics behind this would be.

Maybe something you could try is to run `xinput` to find the ID of the mouse and then run `watch -n 1 xinput list-props 12` and see if the acceleration or transformation matrix changes when you move focus away (of course easier to do if you have two screens).

**Reply by JustCauseWhyNot** [reply]
Thank you very much for the helpful response. I've kinda dropped playing fnv right now, but if I go back to playing it I'll try that. I ended up "fixing" the issue by setting desktop emulation in wine, but that's not a perfect fix since it's not just fallout the game I'm running. But also mod organizer 2, and xEdit which then run as one window instead of dwm managing each window.

**Reply by JustCauseWhyNot** [reply]
Actually would there be a way for dwm to prevent the mouse being moved in unfocused windows? Most games I've tried don't pause when I change workspace tags. I like that they don't pause, but it's annoying moving my mouse resulting in input in the game that I wasn't intending to do.

**Reply by bakkeby** [reply]
To be honest I don't think there is a way to prevent that. It should really be up to the program (game) itself to stop taking input after losing focus. It may be that in this particular case the program is not aware that it has lost focus. That said if it works in gnome or most other window managers then dwm does not unmap windows when they are hidden - they are just being moved to a negative x position, so what might help in this context is the https://dwm.suckless.org/patches/windowmap/ patch.

**Reply by JustCauseWhyNot** [reply]
I appreciate your help. I'll check that patch out. Thank you.

**Reply by JustCauseWhyNot** [reply]
Ok I've applied windowmap, and it doesn't seem to be fixing the issue of changing window focus moving mouse, and then upon refocusing the mouse has inherited where the mouse position was on other workspace.

**Reply by bakkeby** [reply]
Well it was worth a shot. The idea was that the program may behave differently when the window is unmapped, but in this case it doesn't appear to have had any positive effect.

**Reply by JustCauseWhyNot** [reply]
It may for different programs than what I'm trying right now. I agree it was worth a shot thx for letting me know about it.
